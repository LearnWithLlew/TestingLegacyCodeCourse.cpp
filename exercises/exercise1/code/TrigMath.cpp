#include "TrigMath.h"

double TrigMath::SinRaw(int idx) {
    return SIN_TABLE[idx & SIN_MASK];
}

double TrigMath::CosRaw(int idx) {
    return SIN_TABLE[(idx + COS_OFFSET) & SIN_MASK];
}

double TrigMath::Mxatan(double arg) {
    auto argsq = arg * arg;
    auto value = (((p4 * argsq + p3) * argsq + p2) * argsq + p1) * argsq + p0;
    value /= ((((argsq + q4) * argsq + q3) * argsq + q2) * argsq + q1) * argsq + q0;
    return value * arg;
}

double TrigMath::Msatan(double arg) {
    if (arg < sq2m1)
        return Mxatan(arg);
    if (arg > sq2p1)
        return HALF_PI - Mxatan(1 / arg);
    return HALF_PI / 2 + Mxatan((arg - 1) / (arg + 1));
}

void TrigMath::CreateSinTable() {
    for (auto i = 0; i < SIN_SIZE; i++)
        SIN_TABLE[i] = sin(i * TWO_PI / SIN_SIZE);
}

TrigMath::TrigMath() {
    CreateSinTable();
}

double TrigMath::Sin(double angle) {
    return SinRaw(Floor(angle * SIN_CONVERSION_FACTOR));
}

int TrigMath::Floor(double a) {
    return (int) a;
}

double TrigMath::Cos(double angle) {
    return CosRaw(Floor(angle * SIN_CONVERSION_FACTOR));
}

double TrigMath::Tan(double angle) {
    auto idx = Floor(angle * SIN_CONVERSION_FACTOR);
    return SinRaw(idx) / CosRaw(idx);
}

double TrigMath::Sec(double angle) {
    return 1 / Cos(angle);
}

double TrigMath::Cot(double angle) {
    auto idx = Floor(angle * SIN_CONVERSION_FACTOR);
    return CosRaw(idx) / SinRaw(idx);
}

double TrigMath::Asin(double value) {
    if (value > 1)
        return NAN;
    if (value < 0)
        return -Asin(-value);
    auto temp = sqrt(1 - value * value);
    if (value > 0.7)
        return HALF_PI - Msatan(temp / value);
    return Msatan(value / temp);
}

double TrigMath::Acos(double value) {
    if (value > 1 || value < -1)
        return NAN;
    return HALF_PI - Asin(value);
}

double TrigMath::Atan(double value) {
    if (value > 0)
        return Msatan(value);
    return -Msatan(-value);
}

double TrigMath::Atan2(double y, double x) {
    if (y + x == y)
        return y >= 0 ? HALF_PI : -HALF_PI;
    y = Atan(y / x);
    if (x < 0) {
        if (y <= 0) {
            return y + PI;
        } else {
            return y - PI;
        }
    }
    return y;
}

double TrigMath::Acsc(double value) {
    if (value == 0)
        return NAN;
    return Asin(1 / value);
}

double TrigMath::Asec(double value) {
    if (value == 0)
        return NAN;
    return Acos(1 / value);
}

double TrigMath::Acot(double value) {
    if (value == 0)
        return NAN;
    if (value > 0)
        return Atan(1 / value);
    return Atan(1 / value) + PI;
}

