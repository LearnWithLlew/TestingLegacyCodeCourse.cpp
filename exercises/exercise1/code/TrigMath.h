//
// Created by LLEWELLYN FALCO on 12/10/17.
//

#ifndef CATCHPLAYGROUND_TRIGMATH_H
#define CATCHPLAYGROUND_TRIGMATH_H

#define _USE_MATH_DEFINES
#include <math.h>
#include <cmath>

class TrigMath {
public:
    const double PI = M_PI;
    const double SQUARED_PI = PI * PI;
    const double HALF_PI = PI / 2;
    const double QUARTER_PI = HALF_PI / 2;
    const double TWO_PI = 2 * PI;
    const double THREE_PI_HALVES = TWO_PI - HALF_PI;
    const double DEG_TO_RAD = PI / 180;
    const double HALF_DEG_TO_RAD = PI / 360;
    const double RAD_TO_DEG = 180 / PI;
    double SQRT_OF_TWO = std::sqrt(2);
    double HALF_SQRT_OF_TWO = SQRT_OF_TWO / 2;
private  :

    const int SIN_SIZE = 100000;
    double SIN_TABLE[100000];
    const int SIN_MASK = SIN_SIZE - 1;
    const double SIN_CONVERSION_FACTOR = SIN_SIZE / TWO_PI;

    const int COS_OFFSET = SIN_SIZE / 4;

//Arc trig
    const double sq2p1 = 2.414213562373095048802;

    const double sq2m1 = 0.414213562373095048802;
    const double p4 = 0.161536412982230228262E2;
    const double p3 = 0.26842548195503973794141E3;
    const double p2 = 0.11530293515404850115428136E4;
    const double p1 = 0.178040631643319697105464587E4;
    const double p0 = 0.89678597403663861959987488E3;
    const double q4 = 0.5895697050844462222791E2;
    const double q3 = 0.536265374031215315104235E3;
    const double q2 = 0.16667838148816337184521798E4;
    const double q1 = 0.207933497444540981287275926E4;
    const double q0 = 0.89678597403663861962481162E3;

    double sinRaw(int idx);

    double cosRaw(int idx);

    double mxatan(double arg);

    double msatan(double arg);

    void createSinTable();

public:

    TrigMath();


    double sin(double angle);

    int floor(double a);


    double cos(double angle);

    double tan(double angle);

    double csc(double angle) {
        return 1 / sin(angle);
    }

    double sec(double angle);


    double cot(double angle);

    double asin(double value);

    double acos(double value);

    double atan(double value);

    double atan2(double y, double x);

    double acsc(double value);

    double asec(double value);

    double acot(double value);


};

#endif //CATCHPLAYGROUND_TRIGMATH_H
