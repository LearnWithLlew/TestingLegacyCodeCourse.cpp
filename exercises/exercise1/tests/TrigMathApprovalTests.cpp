#include <catch2/catch.hpp>
#include "ApprovalTests.hpp"
#include "TrigMath.h"

#include <utility>

using namespace ApprovalTests;

// See https://approvaltestscpp.readthedocs.io/en/latest/
typedef double (TrigMath::*TrigMathMemFn)(double value);
#define CALL_MEMBER_FN(object,ptrToMember)  ((object).*(ptrToMember))

using Name = const char*;
using NameAndFunction = std::pair<Name, TrigMathMemFn>;
using NamesAndFunctions = std::vector<NameAndFunction>;

std::ostream &operator<<(std::ostream &out,
                         const NameAndFunction &name_and_function) {
  out << "&TrigMath::" + std::string(name_and_function.first);
  return out;
}

TEST_CASE("Demo combinations with pointer to member functions")
{
    TrigMath math;
    // Functions given in the same order as in the TrigMath.h header
    // (and copied in via column selection)
    const NamesAndFunctions functions{
        {"Sin", &TrigMath::Sin},
        {"Cos", &TrigMath::Cos},
        {"Tan", &TrigMath::Tan},
        {"Csc", &TrigMath::Csc},
        {"Sec", &TrigMath::Sec},
        {"Cot", &TrigMath::Cot},
        {"Asin", &TrigMath::Asin},
        {"Acos", &TrigMath::Acos},
        {"Atan", &TrigMath::Atan},
        // We can't test Atan2 this way as it takes 2 arguments
        //{"Atan2", &TrigMath::Atan2},
        {"Acsc", &TrigMath::Acsc},
        {"Asec", &TrigMath::Asec},
        {"Acot", &TrigMath::Acot}
    };

    const std::vector<double> values{1.0, 3.14, -0.1, 0, 0.2, 0.8};

    CombinationApprovals::verifyAllCombinations(
        [&math](auto &name_and_function, auto &value) {
          return CALL_MEMBER_FN(math, name_and_function.second)(value);
        },
        functions, values);
}
