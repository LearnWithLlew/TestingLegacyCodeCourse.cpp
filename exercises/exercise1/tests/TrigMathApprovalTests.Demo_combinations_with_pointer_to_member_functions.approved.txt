(&TrigMath::Sin, 1) => 0.097045
(&TrigMath::Sin, 3.14) => 0.866976
(&TrigMath::Sin, -0.1) => -0.0979798
(&TrigMath::Sin, 0) => 0
(&TrigMath::Sin, 0.2) => 0.065234
(&TrigMath::Sin, 0.8) => 0.00980132
(&TrigMath::Cos, 1) => 0.828843
(&TrigMath::Cos, 3.14) => -0.866633
(&TrigMath::Cos, -0.1) => 0.033231
(&TrigMath::Cos, 0) => 0.00854478
(&TrigMath::Cos, 0.2) => 0.0977954
(&TrigMath::Cos, 0.8) => 0.867602
(&TrigMath::Tan, 1) => 0.117085
(&TrigMath::Tan, 3.14) => -1.0004
(&TrigMath::Tan, -0.1) => -2.94845
(&TrigMath::Tan, 0) => 0
(&TrigMath::Tan, 0.2) => 0.667046
(&TrigMath::Tan, 0.8) => 0.011297
(&TrigMath::Csc, 1) => 10.3045
(&TrigMath::Csc, 3.14) => 1.15343
(&TrigMath::Csc, -0.1) => -10.2062
(&TrigMath::Csc, 0) => inf
(&TrigMath::Csc, 0.2) => 15.3294
(&TrigMath::Csc, 0.8) => 102.027
(&TrigMath::Sec, 1) => 1.2065
(&TrigMath::Sec, 3.14) => -1.15389
(&TrigMath::Sec, -0.1) => 30.0924
(&TrigMath::Sec, 0) => 117.031
(&TrigMath::Sec, 0.2) => 10.2254
(&TrigMath::Sec, 0.8) => 1.1526
(&TrigMath::Cot, 1) => 8.54081
(&TrigMath::Cot, 3.14) => -0.999604
(&TrigMath::Cot, -0.1) => -0.339161
(&TrigMath::Cot, 0) => inf
(&TrigMath::Cot, 0.2) => 1.49915
(&TrigMath::Cot, 0.8) => 88.5188
(&TrigMath::Asin, 1) => 1.57075
(&TrigMath::Asin, 3.14) => nan
(&TrigMath::Asin, -0.1) => -0.100167
(&TrigMath::Asin, 0) => 0
(&TrigMath::Asin, 0.2) => 0.201358
(&TrigMath::Asin, 0.8) => 0.927272
(&TrigMath::Acos, 1) => 0
(&TrigMath::Acos, 3.14) => nan
(&TrigMath::Acos, -0.1) => 1.67092
(&TrigMath::Acos, 0) => 1.57075
(&TrigMath::Acos, 0.2) => 1.36939
(&TrigMath::Acos, 0.8) => 0.643478
(&TrigMath::Atan, 1) => 0.785375
(&TrigMath::Atan, 3.14) => 1.26243
(&TrigMath::Atan, -0.1) => -0.0996687
(&TrigMath::Atan, 0) => 0
(&TrigMath::Atan, 0.2) => 0.197396
(&TrigMath::Atan, 0.8) => 0.674718
(&TrigMath::Acsc, 1) => 1.57075
(&TrigMath::Acsc, 3.14) => 0.324116
(&TrigMath::Acsc, -0.1) => nan
(&TrigMath::Acsc, 0) => nan
(&TrigMath::Acsc, 0.2) => nan
(&TrigMath::Acsc, 0.8) => nan
(&TrigMath::Asec, 1) => 0
(&TrigMath::Asec, 3.14) => 1.24663
(&TrigMath::Asec, -0.1) => nan
(&TrigMath::Asec, 0) => nan
(&TrigMath::Asec, 0.2) => nan
(&TrigMath::Asec, 0.8) => nan
(&TrigMath::Acot, 1) => 0.785375
(&TrigMath::Acot, 3.14) => 0.308316
(&TrigMath::Acot, -0.1) => 1.67042
(&TrigMath::Acot, 0) => nan
(&TrigMath::Acot, 0.2) => 1.37335
(&TrigMath::Acot, 0.8) => 0.896032

